### DISCUSSION TITLE
How to make client-side validation with default values?

### DISCUSSION THREAD

**Discussion by whalemare** [sinclairzx81/typebox#275] [Q&A] [answered]
I have some `UserEntity` with `id` and `gender` and want to validate raw `json` objects received from backend to ensure it correctness. If some field is not exist, I want to use `default` values instead incorrect data. How I can achieve this behavior with type box?

```ts
enum GenderType {
  FEMALE = 'female',
  MALE = 'male',
  UNKNOWN = 'unknown',
}

const UserSchema = Type.Object({
  id: Type.Number(),
  gender: Type.Enum(GenderType, { default: GenderType.UNKNOWN }),
})
```

Something like that when correct value:

```tsx
const result = validate(UserSchema, { id: 1, gender: 'incorrect' })
console.log(result) // { id: 1, gender: "unknown" }


const result = validate(UserSchema, { id: 2, gender: null })
console.log(result) // { id: 1, gender: "unknown" }
```

Something like that when value with error:
```tsx
const result = validate(UserSchema, { id: "string instead number", gender: 'male' })
// undefined | null | ValueError | throw Exception
```

**Comment by sinclairzx81** [comment, marked as answer]
@whalemare Hi, I have actually just pushed a patch / fix for this on `@sinclair/typebox@0.25.3` so information below relates to this version (and above)

### Validation & Default Values

TypeBox doesn't provide functionality to create default values during validation (either the data is correct or it's not). But if you are using Ajv, you can use [useDefaults](https://ajv.js.org/options.html#usedefaults) to automatically set `default` values if certain properties happen to be missing. This will need to be configured on the AJV object. Refer to the Ajv documentation for more information. 

### Value.Create() & Value.Cast()

As noted, while TypeBox doesn't provide a mechanism to automatically create default values for schemas specifically on validate, it does however provide a mechanism to transform a value into something that matches the expected type. See examples below

```typescript
import { Value } from '@sinclair/typebox/value'

enum GenderType {
  FEMALE = 'female',
  MALE = 'male',
  UNKNOWN = 'unknown',
}

const UserSchema = Type.Object({
  id: Type.Number(),
  gender: Type.Enum(GenderType, { default: GenderType.UNKNOWN }),
})

const user0 = Value.Create(UserSchema) 
// > { id: 0, gender: 'unknown' }

const user1 = Value.Cast(UserSchema, { id: 42 }) 
// > { id: 42, gender: 'unknown' })

const user2 = Value.Cast(UserSchema, { gender: 'female' }) 
// > { id: 0, gender: 'female' })
```
The `Value.Cast()` function may be the closest approximation TypeBox has to validate + adding default values. The function is relatively low level so you will need to implement the logic to handle cast somewhere within your validate function. The `Value.Cast()` function is loosely described as "Similar to Create, but retains as much information from the original value as possible. The returned value will match the expected type" so can be used as an alternative to validation (as the returned value is generally always going to be correct)

If building client / server functionality, it is recommended to use `Value.Cast()` on the client before submitting to a server validated method (which is generally faster as both Value.Create() and Value.Cast() are a bit expensive to run on the server side)

Hope this helps!
